File: typescript/string-calculator/src/stringCalculator.ts
const DEFAULT_DELIMITERS = [",", "\n"];
const MAX_NUMBER_ALLOWED = 1000;
export function add(input: string): number {
  if (input.length === 0) {
    return 0;
  }

  // 1) Parse optional header: //X\n
  const { body, custom } = parseHeader(input.trim());

  // 2) Build delimiter set (default + optional custom)
  const delimiters = new Set([",", "\n", ...(custom ? [custom] : [])]);

  // 3) Tokenize body using all delimiters (single pass)
  const tokens = tokenize(body, delimiters);

  // 4) Convert to numbers (trimmed, non-empty)
  const numbers = tokens.map(Number);

  if (numbers.some((n) => Number.isNaN(n))) {
    throw new Error(`invalid token(s): ${tokens.join(",")}`);
  }

  // 5) Validate negatives (collect all, then throw once)
  const negatives = numbers.filter((n: number) => n < 0);
  if (negatives.length > 0) {
    throw new Error(
      `negative numbers are not accepted: ${negatives.join(",")}`,
    );
  }

  // 6) Sum while ignoring > 1000
  return numbers.reduce(
    (acc: number, n: number) => (n > MAX_NUMBER_ALLOWED ? acc : acc + n),
    0,
  );
}

function parseHeader(input: string): { body: string; custom?: string } {
  const str = input.trim();

  if (str.startsWith("//") && str.length >= 4 && str[3] === "\n") {
    return { body: str.slice(4), custom: str[2] };
  }

  return { body: str };
}

function tokenize(body: string, delimiters: Set<string>): string[] {
  const tokens: string[] = [];
  let buf = "";

  for (const ch of body) {
    if (delimiters.has(ch)) {
      if (buf.length) tokens.push(buf.trim());
      buf = "";
    } else {
      buf += ch;
    }
  }
  if (buf.length) tokens.push(buf.trim());
  return tokens.filter(Boolean);
}

File: csharp/StringCalculator/src/StringCalculator/StringCalculator.cs
using System.Text;

namespace StringCalculator;

public class StringCalculator
{
    private const int UpperLimit = 1000;
    private const int HeaderLength = 4;

    public int Add(string input)
    {
        if (input.Length == 0)
        {
            return 0;
        }

        // parse header
        Header parsedHeader = ParseHeader(input.Trim());
        // build delimter
        HashSet<char> delimiter = BuildDelimiters(parsedHeader);
        // tokenize
        List<string> tokens = Tokenize(parsedHeader.body, delimiter);
        // numerize
        int numbers = Numerize(tokens);

        return numbers;
    }

    private static Header ParseHeader(string header)
    {
        if (header.StartsWith("//") && header.Length >= HeaderLength && header[3] == '\n')
        {
            return new Header(body: header.Substring(4), custom: header[2]);
        }

        return new Header(body: header, custom: null);
    }

    private static HashSet<char> BuildDelimiters(Header header)
    {
        HashSet<char> delimiters = new HashSet<char> { ',', '\n' };
        if (header.custom is char c)
        {
            delimiters.Add(c);
        }

        return delimiters;
    }

    private static List<string> Tokenize(string body, HashSet<char> delimiters)
    {
        List<string> tokens = new List<string>();
        StringBuilder buffer = new StringBuilder();

        foreach (var c in body)
        {
            if (delimiters.Contains(c))
            {
                if (buffer.Length > 0)
                {
                    tokens.Add(buffer.ToString().Trim());
                    buffer.Clear();
                }
            }
            else
            {
                buffer.Append(c);
            }
        }

        if (buffer.Length > 0)
        {
            tokens.Add(buffer.ToString().Trim());
        }

        return tokens;
    }

    private static int Numerize(List<string> tokens)
    {
        List<int> negatives = new List<int>();
        int sum = 0;

        foreach (string token in tokens)
        {
            if (!int.TryParse(token, out int num))
            {
                throw new ArgumentException($"Invalid token: '{token}'");
            }

            if (num < 0)
            {
                negatives.Add(num);
                continue;
            }

            if (num <= UpperLimit)
            {
                sum += num;
            }
        }

        if (negatives.Count > 0)
        {
            throw new ArgumentException($"Negative numbers are not allowed: {string.Join(",", negatives)}");
        }

        return sum;
    }
}

public record Header(string body, char? custom) { }
